src/proof/fischlin.rs

```diff
@@ -0,0 +1,357 @@
+            })
+            .collect();
+
+        if hash_checks.iter().any(|res| res.is_err()) {
+            return Err(Error::InternalError);
+        } else if hash_checks
+            .into_iter()
+            .map(|res| res.unwrap())
+            .any(|hash_starts_with_b_zeros| !hash_starts_with_b_zeros)
+        {
+            return Err(proof::Error::ProofVerification)?;
+        }
+
+        self.maurer_proof.verify_inner(
+            self.challenges.map(|challenge| vec![challenge]),
+            language_public_parameters,
+            vec![statement],
+        )
+    }
+
+    /// Compute $common-h$ as a full hash with output length $2\cdot\kappa_c$.
+    fn compute_common_hash(
+        protocol_context: &ProtocolContext,
+        language_public_parameters: &Language::PublicParameters,
+        statement: &Language::StatementSpaceGroupElement,
+        statement_masks_values: &[group::Value<Language::StatementSpaceGroupElement>; REPETITIONS],
+    ) -> Result<<ComputationalSecuritySizedNumber as Concat>::Output> {
+        let mut transcript =
+            super::Proof::<REPETITIONS, Language, ProtocolContext>::setup_transcript(
+                protocol_context,
+                language_public_parameters,
+                vec![statement.value()],
+                statement_masks_values,
+            )?;
+
+        Ok(transcript.challenge(b"common hash"))
+    }
+
+    /// Performs the Fischlin transformation check that $h_i = H(common-h, i, e_i, z_i)$ starts with $b$ zeros.
+    fn hash_starts_with_b_zeros(
+        common_hash: <ComputationalSecuritySizedNumber as Concat>::Output,
+        i: usize,
+        challenge: ComputationalSecuritySizedNumber,
+        response: &WitnessSpaceValue<REPETITIONS, Language>,
+    ) -> Result<bool> {
+        // Set up the transcript
+        let mut transcript = Transcript::new(b"Fischlin hash check");
+        // Add the common hash $common-h$
+        transcript.append_uint(b"common hash", &common_hash);
+        // Add the response index $i$.
+        transcript.append_message(b"i", &i.to_be_bytes());
+        // Add the challenge $e_i$
+        transcript.append_uint(b"challenge", &challenge);
+        // Add the response $z_i$
+        let res = transcript.serialize_to_transcript_as_json(b"response", &response);
+        if res.is_err() {
+            return Err(res.err().unwrap())?;
+        }
+        // Finalize the hash computation to get $h_i = H(common-h, i, e_i, z_i)$.
+        let hash: ComputationalSecuritySizedNumber = transcript.challenge(b"hash");
+
+        // TODO: better name?
+        // Compute $b$ such that $b \cdot \rho >= \kappa$
+        let hash_zero_bits =
+            (ComputationalSecuritySizedNumber::BITS + (REPETITIONS - 1)) / REPETITIONS;
+
+        // TODO: verify correctness
+        Ok(
+            (hash & ComputationalSecuritySizedNumber::from((1u64 << hash_zero_bits) - 1))
```

Comment:
It's clearly true unless your prover does O(2^64) work, but you may add a check that `2^64 > b`.
Given that, yes this works.

Alternatively, doesn't this work?
```suggestion
            (hash & ((ComputationalSecuritySizedNumber::from(1u64) << hash_zero_bits) - 1))
```